Dynamic skill extraction + minor refactoring

joriscram opened the pull request:
In my task to review 440 graduate resumes I came across this repo and slightly adjusted this great initiative, some of it might be useful for others hence this pull request:

- Main functionality of this pull request is that one can add extraction fields in the config and the app will dynamically extract them from the resumes with string matching. (so no coding involved)

in addition:
- Removed spacy dependencies. NLP models can be imported directly. So you do not need the assets folder and installation (especially on windows machines) is more convenient without spacy.
- My dataset contained anonymous resumes, so had to escape for this situation in the name extraction
- extract function does not need nlp model. In addition utf-8 encoding is ensured by a function and not by forcing the system default encodings.

joriscram committed:
- removed spacy dependency since only importable nlp models are used
- removed unicode enforcing and solved encoding in the application
- extracting skills refactored such you don ont have to make code adjustments when adding new skill categories or skills
- renamed confs to config

joriscram committed:
modified temmplate structure such you can add exctractors and skills without touching the code

bjherger commented on bin/main.py:
```diff
@@ -8,29 +8,26 @@
-reload(sys)
-sys.setdefaultencoding('utf8')
-
+import en_core_web_sm
```
This import isn't in the `environment.yaml` file. Could we update that file to include it?

`import en_core_web_md`

And change `environment.yaml` to read: 
```
name: resume
channels:
- !!python/unicode
  'defaults'
dependencies:
- !!python/unicode
  'openssl=1.0.2l=0'
- !!python/unicode
  'pip=8.1.2=py27_0'
- !!python/unicode
  'python=2.7.12=1'
- !!python/unicode
  'readline=6.2=2'
- !!python/unicode
  'setuptools=26.1.1=py27_0'
- !!python/unicode
  'sqlite=3.13.0=0'
- !!python/unicode
  'tk=8.5.18=0'
- !!python/unicode
  'wheel=0.29.0=py27_0'
- !!python/unicode
  'zlib=1.2.8=3'
- pip:
  - argcomplete==1.8.2
  - beautifulsoup4==4.5.3
  - boto==2.48.0
  - bz2file==0.98
  - certifi==2017.7.27.1
  - chardet==3.0.4
  - cymem==1.31.2
  - cytoolz==0.8.2
  - dill==0.2.7.1
  - docx2txt==0.6
  - ebooklib==0.15
  - en-core-web-md==1.2.0
  - ftfy==4.4.3
  - gensim==3.0.1
  - html5lib==0.999999999
  - idna==2.6
  - lxml==4.1.0
  - murmurhash==0.26.4
  - numpy==1.13.3
  - olefile==0.44
  - pandas==0.20.3
  - pathlib==1.0.1
  - pillow==4.3.0
  - plac==0.9.6
  - pocketsphinx==0.1.3
  - preshed==1.0.0
  - python-dateutil==2.6.1
  - python-pptx==0.6.5
  - pytz==2017.2
  - pyyaml==3.12
  - regex==2017.9.23
  - requests==2.18.4
  - scipy==0.19.1
  - six==1.10.0
  - smart-open==1.5.3
  - spacy==1.9.0
  - speechrecognition==3.6.3
  - termcolor==1.1.0
  - textract==1.6.1
  - thinc==6.5.2
  - toolz==0.8.2
  - tqdm==4.19.4
  - ujson==1.35
  - urllib3==1.22
  - wcwidth==0.1.7
  - webencodings==0.5.1
  - wrapt==1.10.11
  - xlrd==1.0.0
  - xlsxwriter==1.0.2
```

joriscram replied:
I will revert this since the en_core_web_sm is not available in the conda or Pypi repository and can only be dowloaded via spacy (with the command: 'python -m spacy download en')

bjherger commented on bin/main.py:
```diff
@@ -8,29 +8,26 @@
     :rtype: None
     """
-    logging.getLogger().setLevel(logging.DEBUG)
+    logging.getLogger().setLevel(logging.INFO)
```
Good call

bjherger commented on bin/main.py:
```diff
@@ -41,6 +38,12 @@ def main():
+    try:
+        return unicode(textract.process(f), "utf-8")
+    except UnicodeDecodeError, e:
+        return e
```
Could we return the empty string instead? Returning the error message might make it appear that the person's resume has error text.

How about:
```
except UnicodeDecodeError, e:
        return ''
```

bjherger reviewed (commented):
This looks awesome! Thanks for the assist. 

I have one concern around adding the spacy import to the `environment.yml` file, and a few other nits. Otherwise, this looks great!

bjherger commented:
@joriscram : PR review provided. Please respond to the feedback. 

After one iteration, we should be able to merge this in. Thanks for your input!

joriscram committed:
implemented review comments
- also changed the readme with the instruction to install a spacy nlp model, this way you do not need the assets folder.

bjherger commented:
@joriscram Thanks for this awesome contribution. I've validated that the new functionality and spacy asset loading works, and I'm happy to merge this in.

bjherger commented:
+1, LGTM. Merging in w/ +1s from myself and @joriscram. No open requests or comments on PR.
